File: LeetCode/451-sort-characters-by-frequency/ex_451.py
class Solution(object):
    def frequencySort(self, s):
        """
        :type s: str
        :rtype: str
        """
        heap = []
        res = []
        char_count = collections.Counter(s)
        for c in char_count:
            t = (-char_count[c], c)
            for i in xrange(-t[0]):
                heapq.heappush(heap, t)
        while heap:
            res.append(heapq.heappop(heap)[1])
        return "".join(res)

File: LeetCode/451-sort-characters-by-frequency/ex_451.cpp
class Solution {
public:
    string frequencySort(string s) {
        priority_queue<pair<int, char>> pq;
        unordered_map<char, int> char_count = countChar(s);
        for (const auto &p: char_count) {
            for (int i = 0; i != p.second; ++i) {
                pq.push(make_pair(p.second, p.first));
            }
        }
        string res;
        while (!pq.empty()) {
            auto p = pq.top();
            res += p.second;
            pq.pop();
        }
        return res;
    }
    
private:
    unordered_map<char, int> countChar(string &s) {
        unordered_map<char, int> res;
        for (const char c : s)  res[c]++;
        return res;
    }

};